refactor: Number field-editor [BAU-722]

bgutsol opened the pull request:
- Use type="text" instead of type="number", [see why](https://github.com/contentful/field-editors/blob/ea23844c6d4ef2cbe0f9616a36b219a00b812e09/packages/number/src/NumberEditor.tsx#L120-L122)

- Number field editor does not allow entering nun-numeric input, like a native input type="number"
- Number field editor has step buttons, like a native input type="number"
- Arrow keys can be used to step as well, like a native input type="number"
- Don't use `inputmode=decimal/numeric` because some keyboards on mobile devices don't allow typing minus `-` or dot `.`. For example, Safari on IOS.


## Related PRs:
- https://github.com/contentful/field-editors/pull/1020
- https://github.com/contentful/field-editors/pull/1112

massao commented on packages/number/package.json:
```diff
@@ -24,9 +24,7 @@
     "@contentful/field-editor-shared": "^1.1.3",
-    "emotion": "^10.0.17",
-    "lodash": "^4.17.15",
-    "lodash-es": "^4.17.15"
```
🥇

massao reviewed (approved):
LGTM, should we consider adding this to forma??

bgutsol commented:
> LGTM, should we consider adding this to forma??

Yes 💯

mshaaban0 reviewed (approved):
LGTM. Not sure about the version release, is this a breaking change?

giotiskl commented on packages/number/src/parseNumber.spec.ts:
```diff
@@ -1,16 +1,56 @@
+});
+
+describe('isNumberInputValueValid', () => {
+  it('should correctrly validate input when type is Integer', () => {
```
```suggestion
  it('should correctly validate input when type is Integer', () => {
```

giotiskl commented on packages/number/src/NumberEditor.tsx:
```diff
@@ -20,73 +25,144 @@ export interface NumberEditorProps {
+  };
+
+  const changeValueByStep = (type: 'increment' | 'decrement') => {
+    const currentValue = isNaN(+inputValue) ? 0 : +inputValue;
```
```suggestion
    const currentValue = Number.isNaN(+inputValue) ? 0 : +inputValue;
```
we should use `Number.isNaN` over `isNaN`, explanation here: https://stackoverflow.com/questions/33164725/confusion-between-isnan-and-number-isnan-in-javascript

giotiskl commented on packages/number/src/NumberEditor.tsx:
```diff
@@ -20,73 +25,144 @@ export interface NumberEditorProps {
+    if (stringSdkValue !== inputValue) {
+      setInputValue(stringSdkValue);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- we want to trigger it only when sdkValue had been changed
```
```suggestion
    // eslint-disable-next-line react-hooks/exhaustive-deps -- we want to trigger it only when sdkValue has changed
```

giotiskl commented on packages/number/src/NumberEditor.tsx:
```diff
@@ -20,73 +25,144 @@ export interface NumberEditorProps {
+    }
+  };
+
+  const changeValueByStep = (type: 'increment' | 'decrement') => {
```
instead of hardcoding strings in a union value I'd propose introducing an enum to avoid magic string usage:
```
enum StepChangeType {
  Increment = 'increment',
  Decrement = 'decrement',
}
```

```suggestion
  const changeValueByStep = (type: StepChangeType) => {
```

giotiskl commented on packages/number/src/NumberEditor.tsx:
```diff
@@ -20,73 +25,144 @@ export interface NumberEditorProps {
+    const keyToFnMap: {
+      [key: string]: () => void;
+    } = {
+      ArrowUp: () => changeValueByStep('increment'),
```
this would then become:
```suggestion
      ArrowUp: () => changeValueByStep(StepChangeType.Increment),
```
if we introduce the enum, similarly for the item below

giotiskl commented on packages/number/src/parseNumber.ts:
```diff
@@ -1,51 +1,16 @@
-      value: undefined,
-    };
+export function parseNumber(value: string, type: string) {
+  if (!value || isNaN(+value)) {
```
```suggestion
  if (!value || Number.isNaN(+value)) {
```

giotiskl reviewed (changes requested):
Looks good overall, my main suggestion would be to rely on `Number.isNaN` instead of `isNaN`, explanation in the comments.

bgutsol commented:
> Looks good overall, my main suggestion would be to rely on `Number.isNaN` instead of `isNaN`, explanation in the comments.

@giotiskl thanks for reviewing! I've addressed your comments, please take a look

denkristoffer commented on packages/number/src/NumberEditor.tsx:
```diff
@@ -20,73 +24,150 @@ export interface NumberEditorProps {
+        className={styles.input}
+        min={range.min}
+        max={range.max}
+        step={NUMBER_STEP}
```
Does this work with text input?

bgutsol replied:
Yes, it's actually the same role that input type="number" has: https://www.digitala11y.com/spinbutton-role/

denkristoffer replied:
I meant the `step` attribute. It doesn't seem to be valid: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/step

bgutsol replied:
step can be removed

denkristoffer reviewed (commented):
What does it look like?

bgutsol commented:
> What does it look like?

@denkristoffer you can take a look on preview https://deploy-preview-1203--contentful-field-editors.netlify.app/number

denkristoffer commented:
Looks good, how about some hover and active styling on the buttons though?

bgutsol commented:
> Looks good, how about some hover and active styling on the buttons though?

Active styling is in place. With hover I agree, we can add some

denkristoffer commented:
> Active styling is in place. With hover I agree, we can add some

I mean the clicked state, I'm not seeing it:

https://user-images.githubusercontent.com/219017/185390585-683f38c2-d25d-479c-bb47-641fa9960917.mov

bgutsol commented:
> > Active styling is in place. With hover I agree, we can add some
> 
> I mean the clicked state, I'm not seeing it:
> 
>  Screen.Recording.2022-08-18.at.14.04.46.mov

The current active style doesn't work in firefox for some reason 🤔

bgutsol commented:
> > > Active styling is in place. With hover I agree, we can add some
> > 
> > 
> > I mean the clicked state, I'm not seeing it:
> > Screen.Recording.2022-08-18.at.14.04.46.mov
> 
> The current active style doesn't work in firefox for some reason 🤔

Unfortunately, there is no easy fix for it, it's just the way [Firefox implemented its core logic](https://bugzilla.mozilla.org/show_bug.cgi?id=771241), and it's different from the other browsers. 
So basically `event.preventDefault();`, that I call within `onPointerDown` callback, cancels the active state for our buttons (only in Firefox). But I can't remove the `event.preventDefault();` because I need it to switch the focus from the buttons back to the input for a11y purposes. So the only way to fix it: is by setting our custom `.active` class. But it doesn't worth the effort. What's your thoughts @denkristoffer?
